lib/utils.nix

```diff
@@ -125,27 +125,185 @@ rec {
+      string,
+      buildSurround,
+      surroundWithNewlines,
+    }:
+    let
+      findSurround =
+        {
+          string,
+          buildSurround,
+          depth,
+        }:
+        let
+          surround = buildSurround depth;
+        in
+        if lib.hasInfix surround.close string then
+          findSurround {
+            inherit string buildSurround;
+            depth = depth + 1;
+          }
+        else
+          surround;
+
+      surround = findSurround {
+        inherit string buildSurround;
+        depth = 0;
+      };
+      newlines = if surroundWithNewlines then "\n" else "";
+    in
+    "${surround.open}${newlines}${string}${newlines}${surround.close}";
+
+  /**
+    Given a string, escape it as a Lua [long literal].
+
+    [long literal]: https://www.lua.org/manual/5.4/manual.html#3.1
+
+    # Examples
+
+    ```nix
+    nix-repl> escapeStringForLua "simple"
+    "[[simple]]"
+    ```
+
+    ```nix
+    nix-repl> escapeStringForLua "]]"
+    "[=[]]]=]"
+    ```
+
+    # Type
+
+    ```
+    escapeStringForLua :: String -> String
+    ```
+  */
+  escapeStringForLua =
+    string:
+    escapeString {
+      inherit string;
+      surroundWithNewlines = false;
+
+      # [[
+      # [=[
+      # [==[
+      # ...
+      buildSurround =
+        depth:
+        let
+          equals = lib.strings.replicate depth "=";
+        in
+        {
+          open = "[${equals}[";
+          close = "]${equals}]";
+        };
+    };
+
+  /**
+    Given a string, escape it as a Vimscript [:let-heredoc], for use by [:lua].
+
+    [:let-heredoc]: https://neovim.io/doc/user/eval.html#%3Alet-heredoc
+    [:lua]: https://neovim.io/doc/user/lua.html#%3Alua-heredoc
+
+    # Examples
+
+    ```nix
+    escapeStringForVimscript "simple"
+    => "<< EOF\nsimple\nEOF"
+    ```
+
+    ```nix
+    escapeStringForVimscript "EOF"
+    => "<< EOFF\nEOF\nEOFF"
+    ```
+
+    # Type
+
+    ```
+    escapeStringForVimscript :: String -> String
+    ```
+  */
+  escapeStringForVimscript =
+    string:
+    (
+      let
+        escapedString = escapeString {
+          inherit string;
+          surroundWithNewlines = true;
+
+          # EOF
+          # EOFF
+          # EOFFF
+          # ...
+          buildSurround =
+            depth:
+            let
+              fs = lib.strings.replicate (depth + 1) "F";
+              token = "EO${fs}";
+            in
+            {
+              open = token;
+              close = token;
+              inherit depth;
+            };
+        };
+      in
+      "<< ${escapedString}"
+    );
+
   # Wrap Vimscript for using in lua,
   # but only if the string contains something other than whitespaces
-  # TODO: account for a possible ']]' in the string
   wrapVimscriptForLua =
-    string:
-    lib.optionalString (hasContent string) ''
-      vim.cmd([[
-      ${string}
-      ]])
-    '';
+    string: lib.optionalString (hasContent string) "vim.cmd(${escapeStringForLua string})";
```

Comment:
Maybe! Your call. Hopefully none of your users are using these functions directly (I am not).

Since it affects the public API of nixvim, I'd vote for doing that in a separate PR at least, since this has turned into quite a bit of code.

Reply:
Sure, we can tackle that another time. Historically we haven't been _too_ concerned about deprecating or removing things from the public API in cases where it seems unlikely that anyone is using them. But you are correct we ought to be more careful really.

As for `surroundFor` I'm happy either way. Keeping it out of the public API is probably better. I'm also curious about what the implementation would look like if we inlined it into both `buildSurround` implementations. I.e. would the potential simplification outweigh the code duplication. :thinking:

Reply:
> I'm also curious about what the implementation would look like if we inlined it into both buildSurround implementations. I.e. would the potential simplification outweigh the code duplication. 🤔

Good question. The fact that it has such a weird name is perhaps a smell that it shouldn't really exist. That said, I can't really bring myself to duplicate it (perhaps because I sunk all that time into writing a docstring for it :rofl:).

I've floated `surroundFor` up to the top, but I'm leaving this conversation unresolved. I defer to you as maintainer on whatever you think is best.

Reply:
Oh, nevermind! I'm just seeing your latest two comments (https://github.com/nix-community/nixvim/pull/2272#issuecomment-2362184330 and https://github.com/nix-community/nixvim/pull/2272#issuecomment-2362218500), and I'm inclined to agree with you.

Updating the PR. Standby...

Reply:
Updated. I agree that inlining the implementation is better. Love the result, thanks!

Marking this thread resolved.